### PR TITLE
Routing via interface

### PR DESCRIPTION
**Pull request by alexander-yevsyukov**
This PR allows to define routing for a super-interface of messages (see #1037).

It also addresses deprecations recently introduced by Truth.

### CODE REVIEW

**Comment by alexander-yevsyukov** [comment]
@armiol, PTAL.

**Comment by alexander-yevsyukov** [comment]
@armiol, PTAL.

File: server/src/main/java/io/spine/server/route/EventRouting.java [hunk 1]
```diff
@@ -93,36 +94,50 @@ private EventRouting(EventRoute<I, EventMessage> defaultRoute) {
...
-     * Sets a custom route for the passed event class.
+     * Sets a custom route for the passed event type.
      *
      * <p>Such a mapping may be required when...
      * <ul>
-     *     <li>An an event message should be matched to more than one entity (e.g. several
-     *         projections updated in response to one event).
-     *     <li>The type of an event producer ID (stored in the event context) differs from the type
-     *         of entity identifiers ({@code <I>}.
+     *   <li>An an event message should be matched to more than one entity (e.g. several
```
Comment (armiol) [inline server/src/main/java/io/spine/server/route/EventRouting.java]: Let's kill the second `an`.

File: server/src/main/java/io/spine/server/route/MessageRouting.java [hunk 2]
```diff
@@ -68,52 +65,55 @@
...
      *
-     * <p>Such a mapping may be required when...
-     * <ul>
-     *     <li>A message should be matched to more than one entity.
-     *     <li>The type of an message producer ID (stored in the message context) differs from the
-     *         type of entity identifiers.
-     * </ul>
+     * <p>The type can be either a class or interface. If the routing schema already contains an
+     * entry with the same type or a super-interface of the passed type
+     * {@link IllegalStateException} will be thrown.
```
Comment (armiol) [inline server/src/main/java/io/spine/server/route/MessageRouting.java]: Let's add an "an": `an {@link IllegalStateException} will be thrown.`

File: server/src/main/java/io/spine/server/route/MessageRouting.java [hunk 3]
```diff
@@ -68,52 +65,55 @@
...
+            if (match.direct()) {
+                throw newIllegalStateException(
+                        "The route for the message class `%s` already set. " +
+                                "Please remove the route (`%s`) before setting new route.",
+                        requestedClass, entryClass);
+            } else {
+                throw newIllegalStateException(
+                        "The route for the message class `%s` already defined via " +
+                                "the interface `%s`. If you want to have specific routing for " +
+                                "the class `%s` please put it before the routing for " +
```
Comment (armiol) [inline server/src/main/java/io/spine/server/route/MessageRouting.java]: `, please`.

File: server/src/main/java/io/spine/server/route/MessageRouting.java [hunk 4]
```diff
@@ -158,13 +192,49 @@ public void remove(Class<? extends M> messageClass) {
...
+    /**
+     * Provides information on routing availability.
+     */
+    final class Match {
+
+        private final Class<? extends M> requestedClass;
+        private final @Nullable Route<M, C, R> route;
+        private final @Nullable Class<? extends M> entryClass;
+
+        private Match(Class<? extends M> requestedClass,
```
Comment (armiol) [inline server/src/main/java/io/spine/server/route/MessageRouting.java]: Can you please describe the parameters? It's not 100% clear what a `requestedClass` is.

File: build.gradle [hunk 5]
```diff
@@ -257,6 +257,16 @@ subprojects {
...
     apply from: deps.scripts.pmd
+
+    compileJava {
+        options.encoding = 'UTF-8'
+        options.compilerArgs << "-Xlint:unchecked" << "-Xlint:deprecation"
+    }
+
+    compileTestJava {
+        options.encoding = 'UTF-8'
+        options.compilerArgs << "-Xlint:unchecked" << "-Xlint:deprecation"
```
Comment (dmdashenkov) [inline build.gradle]: Please perform this configuration in `config/gradle/javac-args.gradle`. The part with the encoding is already there.

File: testutil-server/src/main/java/io/spine/testing/server/EmittedMessageSubject.java [hunk 6]
```diff
@@ -102,7 +102,7 @@ public final void isNotEmpty() {
     }
 
     private IterableSubject assertActual() {
-        return check().that(actual());
+        return check("actual()").that(actual());
```
Comment (dmdashenkov) [inline testutil-server/src/main/java/io/spine/testing/server/EmittedMessageSubject.java]: This is of a little help to a developer who's not familiar with the `Subject`'s internal wiring.

File: testutil-server/src/main/java/io/spine/testing/server/EmittedMessageSubject.java [hunk 7]
```diff
@@ -143,9 +143,13 @@ public final S withType(Class<? extends M> messageClass) {
         } else {
             List<T> filtered =
                     Streams.stream(actual)
-                           .filter(m -> m.is(messageClass))
+                           .filter(m -> {
+                               @SuppressWarnings({"unchecked", "RedundantSuppression"})
```
Comment (dmdashenkov) [inline testutil-server/src/main/java/io/spine/testing/server/EmittedMessageSubject.java]: Please document this stange suppression.

**Pull request merged**